server/sonar-server/src/main/java/org/sonar/server/component/index/ComponentIndex.java

```diff
@@ -58,36 +62,83 @@ public ComponentIndex(EsClient client, UserSession userSession) {
     this.userSession = userSession;
   }
 
-  public List<String> search(ComponentIndexQuery query) {
+  public List<ComponentsPerQualifier> search(ComponentIndexQuery query) {
+    return search(query, ComponentIndexSearchFeature.values());
+  }
+
+  @VisibleForTesting
+  List<ComponentsPerQualifier> search(ComponentIndexQuery query, ComponentIndexSearchFeature... features) {
+    if (query.getQualifiers().isEmpty()) {
+      return Collections.emptyList();
+    }
+
     SearchRequestBuilder request = getClient()
       .prepareSearch(INDEX_COMPONENTS)
       .setTypes(TYPE_COMPONENT)
-      .setFetchSource(false);
+      .setFetchSource(false)
+
+      // the search hits are part of the aggregations
+      .setSize(0)
+
+      .setQuery(createQuery(query, features));
+
+    query.getQualifiers().stream()
+      .map(q -> createAggregation(query, q))
+      .forEach(request::addAggregation);
+
+    SearchResponse searchResponse = request.get();
+
+    return query.getQualifiers().stream()
+      .flatMap(q -> {
 
-    query.getLimit().ifPresent(request::setSize);
+        InternalFilter agg = searchResponse.getAggregations().get(q);
+        InternalTopHits docs = agg.getAggregations().get("docs");
 
-    request.setQuery(createQuery(query));
+        SearchHits hits = docs.getHits();
 
-    return Arrays.stream(request.get().getHits().hits())
-      .map(SearchHit::getId)
+        long totalHits = hits.totalHits();
+        if (totalHits < 1) {
+          return Stream.empty();
+        }
+
+        List<String> componentUuids = Arrays.stream(hits.getHits()).map(SearchHit::getId)
+          .collect(Collectors.toList());
```

Comment:
should be replaced by `.collect(org.sonar.core.util.stream.Collectors.toList(hits.getHits().length))`